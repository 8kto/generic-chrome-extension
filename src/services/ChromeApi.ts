/**
 * Class encapsulates the Chrome browser API.
 * No direct `chrome.*` calls should be performed from the other scripts.
 */
export default class ChromeApi {
  static async getActiveTabId(): Promise<number> {
    const [tab] = await chrome.tabs.query({ active: true, currentWindow: true })

    return tab.id
  }

  static async reloadTab(tabId: number): Promise<void> {
    await chrome.tabs.reload(tabId)
    window.close()
  }

  static async executeScript<Args extends unknown[], Res = unknown>(
    injection: chrome.scripting.ScriptInjectionCustom<Args>,
    injectionResultsHandler?: chrome.scripting.ScriptInjectionResultsHandler<Res>
  ) {
    return chrome.scripting.executeScript(injection, injectionResultsHandler)
  }

  static getManifest() {
    return chrome.runtime.getManifest()
  }
}
