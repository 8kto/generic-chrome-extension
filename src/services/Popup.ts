import type { VariableUpdatePayload } from 'types'

export default class Popup {
  /**
   * Guess the prefix from the present option and return
   * the list of possible variant names
   */
  static getVariantsOptions(presentOption = ''): string[] {
    const defaultVariationsNumber = 3
    const matched = presentOption && presentOption.match(/^(variation_|v)(\d+)/)

    const getOptions = (prefix: string, num: number) =>
      new Array(num).fill(null).map((_, i) => `${prefix}${i + 1}`)

    const decorateOptionsList = (options: string[]) => {
      const res = ['default', ...options, 'Custom']

      // Keep the custom value (not variationN string)
      if (
        presentOption &&
        !presentOption.match(/^(?:default|(?:variation_|v)\d+)$/)
      ) {
        res.unshift(presentOption)
      }

      return res
    }

    // We cannot guess the correct prefix, so generate all possible
    if (presentOption === 'default' || !matched) {
      return decorateOptionsList([
        ...getOptions('v', 3),
        ...getOptions('variation_', 3),
      ])
    }

    const [, prefix, variantNum] = matched

    return decorateOptionsList(
      getOptions(prefix, Math.max(defaultVariationsNumber, +variantNum))
    )
  }
}
