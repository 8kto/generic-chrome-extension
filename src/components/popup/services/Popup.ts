import ChromeApi from 'services/ChromeApi'

type DetailsTabContentHandler = {
  selector: string
  regexp: RegExp
  handler: (...args: string[]) => string
}

/**
 * Service contains static logic used by `ViewController`
 */
export default class Popup {
  /**
   * Guess the prefix from the present option and return
   * the list of possible variant names
   */
  static getVariantsOptions(presentOption = ''): string[] {
    const defaultVariationsNumber = 3
    const matched = presentOption && presentOption.match(/^(variation_|v)(\d+)/)

    const getOptions = (prefix: string, num: number) =>
      new Array(num).fill(null).map((_, i) => `${prefix}${i + 1}`)

    const decorateOptionsList = (options: string[]) => {
      const res = ['default', ...options, 'Custom']

      // Keep the custom value (not variationN string)
      if (
        presentOption &&
        !presentOption.match(/^(?:default|(?:variation_|v)\d+)$/)
      ) {
        res.unshift(presentOption)
      }

      return res
    }

    // We cannot guess the correct prefix, so generate all possible
    if (presentOption === 'default' || !matched) {
      return decorateOptionsList([
        ...getOptions('v', 3),
        ...getOptions('variation_', 3),
      ])
    }

    const [, prefix, variantNum] = matched

    return decorateOptionsList(
      getOptions(prefix, Math.max(defaultVariationsNumber, +variantNum))
    )
  }

  static updateExtensionVersion(): void {
    const versionContainer = document.getElementById('igel-version')
    if (versionContainer) {
      const manifest = ChromeApi.getManifest()
      versionContainer.innerText = manifest.version
    }
  }

  static updateDetailsTabContent(cookies: string): void {
    const defaultHandler = (v: string) => v
    const containers: DetailsTabContentHandler[] = [
      {
        selector: '#feature-branch-container',
        regexp: /x-featurebranch=([^;$]+)[;$]/,
        handler: defaultHandler,
      },
      {
        selector: '#feature-flag-targeting-params-container',
        regexp: /feature-flag-targeting=([^;$]+)[;$]/,
        handler: (val: string): string => {
          const parsed = JSON.stringify(JSON.parse(val), null, '  ')

          return `<pre>${parsed}</pre>`
        },
      },
    ]

    containers.forEach(def => {
      const container = document.querySelector<HTMLElement>(def.selector)

      if (container) {
        const matched = cookies.match(def.regexp)
        if (matched && matched[1]) {
          container.innerHTML = def.handler(matched[1])
        }
      }
    })
  }
}
