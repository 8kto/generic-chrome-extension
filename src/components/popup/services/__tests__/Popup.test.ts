import ChromeApi from 'services/ChromeApi'

import Popup from './../Popup'

describe('popup service test', () => {
  describe('getVariantsOptions', () => {
    const completeSet = [
      'default',
      'v1',
      'v2',
      'v3',
      'variation_1',
      'variation_2',
      'variation_3',
      'Custom',
    ]

    it('returns all possible options', () => {
      expect(Popup.getVariantsOptions('')).toStrictEqual(completeSet)
    })

    it.each([null, undefined, ''])('handles invalid input', input => {
      // @ts-ignore-next-line
      expect(Popup.getVariantsOptions(input)).toStrictEqual(completeSet)
    })

    it('returns extended options for old names', () => {
      expect(Popup.getVariantsOptions('variation_1')).toStrictEqual([
        'default',
        'variation_1',
        'variation_2',
        'variation_3',
        'Custom',
      ])
    })

    it('returns extended options for new names', () => {
      expect(Popup.getVariantsOptions('v1')).toStrictEqual([
        'default',
        'v1',
        'v2',
        'v3',
        'Custom',
      ])
    })

    it('includes present option', () => {
      expect(Popup.getVariantsOptions('my-var')).toStrictEqual([
        'my-var',
        ...completeSet,
      ])
    })

    it('returns options extended up to the current variant', () => {
      expect(Popup.getVariantsOptions('variation_6')).toStrictEqual([
        'default',
        'variation_1',
        'variation_2',
        'variation_3',
        'variation_4',
        'variation_5',
        'variation_6',
        'Custom',
      ])

      expect(Popup.getVariantsOptions('v6')).toStrictEqual([
        'default',
        'v1',
        'v2',
        'v3',
        'v4',
        'v5',
        'v6',
        'Custom',
      ])
    })
  })

  describe('updateExtensionVersion', () => {
    it('updates the version container', () => {
      document.body.innerHTML = `<p><strong>Version</strong>: <span id="igel-version"></span></p>`
      jest
        .spyOn(ChromeApi, 'getManifest')
        // @ts-ignore
        .mockReturnValue({ version: 'v1.1.1' })

      const versionContainer = document.getElementById('igel-version')

      expect(versionContainer?.innerHTML).toBe('')
      Popup.updateExtensionVersion()
      expect(versionContainer?.innerHTML).toBe('v1.1.1')

      jest.resetAllMocks()
    })
  })

  describe('updateDetailsTabContent', () => {
    it('updates the details', () => {
      document.body.innerHTML =
        `<span id="feature-branch-container">NA</span>` +
        `<span id="feature-flag-targeting-params-container">NA</span>`

      const featureBranchContainer = document.getElementById(
        'feature-branch-container'
      )
      const targetingParamsContainer = document.getElementById(
        'feature-flag-targeting-params-container'
      )

      expect(targetingParamsContainer?.innerHTML).toBe('NA')
      expect(featureBranchContainer?.innerHTML).toBe('NA')

      Popup.updateDetailsTabContent(
        [
          `x-featurebranch=TestBranch`,
          `feature-flag-targeting={"browser-family":"Chrome","browser-version":"101.0.4951.54","device":"mobile","expa":"","isInternalIP":false,"tenant":"com"};`,
        ].join('; ')
      )

      expect(featureBranchContainer?.innerHTML).toBe('TestBranch')
      expect(targetingParamsContainer?.innerHTML).toMatchSnapshot()
    })
  })
})
