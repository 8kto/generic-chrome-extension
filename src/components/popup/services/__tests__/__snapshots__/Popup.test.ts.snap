// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`popup service test updateDetailsTabContent updates the details 1`] = `
"<pre>{
  \\"browser-family\\": \\"Chrome\\",
  \\"browser-version\\": \\"101.0.4951.54\\",
  \\"device\\": \\"mobile\\",
  \\"expa\\": \\"\\",
  \\"isInternalIP\\": false,
  \\"tenant\\": \\"com\\"
}</pre>"
`;
